design/bytecode.md

```diff
@@ -0,0 +1,106 @@
+# Byte code
+
+The Byte code of Seed is generated from the SeedIR code, and executed by the
+SeedInterpreter as a virtual machine. The byte code instructions are 32 bits in
+length. All instructions have an opcode in the first 7 bits, so there are at
+most 128 kinds of opcode. Other fields of the instructions are listed below:
+
+- A (8 bits): The operand A
+- B (8 bits): The operand B
+- C (8 bits): The operand C
+- K (1 bit): Indicates if the operand C is the index of the register (K = 0) or
+  constant (K = 1)
+- Bx (17 bits): The operand Bx, combined with B, C and K
+- sBx (17 bits): The operand sBx, A signed argument in the range of [-Max, Max)
+  (Max = 2 ^ 16). The actual value is the unsigned value (Bx) minus Max, so that
+  -Max is represented by 0, and Max - 1 is represented by 2 \* Max - 1.
+
+The layout of the instructions is illustrated as below:
+
+```text
+
+     3 3 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0
+     1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+    |K|      C (8)    |      B (8)    |     A (8)     |   Op (7)    |
```

Comment:
Per the below design of AddI, I guess a better idea is to have a separate line for sC, e.g.:

| sC(9) | B(8) | A (8) | Op (7) |

So that the flag K in this graph won't need to have two different meanings (a constant flag, and, a sign flag).

Reply:
I found another instruction in Lua 5.4:

OP_EQI,/*	A sB k	if ((R[A] == sB) ~= k) then pc++		*/

So sB and K are used at the same time. Maybe K is not used as a sign flag. and the range of sB and sC should be in [-128, 128), instead of [-256, 256).